C++: Don't generate dataflow nodes for functions with summaries

MathiasVP opened the pull request:
Not generating dataflow nodes for instructions/operands inside summarized callables has a big impact when using STL since the bodies of the instantiations are always in the database. So by excluding these we will avoid generating dataflow nodes in all of the instantiations which may be pretty significant if there are many instantiations.

Additionally, when we have flow through a function both through MaD and through the source code we may unnecessarily hit the field flow branch limit by duplicating flow. Staying below the field flow branch limit means we allow field flow in more functions.

DCA results looks good. Here's my breakdown of them:
### `cpp/path-injection`

We gain 1 new result on `cpp/path-injection`. I've confirmed that this is because we now stay below the default field flow branch limit (i.e., `2`) for flow out of [this call to push_back](https://github.com/openjdk/jdk/blob/8dc43aa0fe8cdba2a2953258de02c6afa072987a/src/jdk.jpackage/share/native/applauncher/AppLauncher.h#L49). Before, we had out flow from both the MaD summary and the source code which resulted in going over the limit. But now we only have the MaD summary-proided out flow which keeps us below the threshold. So, unlike on `main`, field flow is now permitted in the enclosing function.

### `cpp/non-constant-format`

We lose 60 results on SAMATE for this query. They all appear to be false positives that happen because of the generous [`isSource` in the query](https://github.com/github/codeql/blob/4d2ec75ef22ffe8be19fd2a08d4bc2ff16aa5152/cpp/ql/src/Likely%20Bugs/Format/NonConstantFormat.ql#L97) that makes us start flow at some random output parameter of a call to `delete` deep inside the destructor of an iterator inside the libstdc++. Obviously, that's not what the query is supposed to be finding and I doubt that any of our queries will benefit from starting flow deep inside the implementation of a MaD summarized function.

jketema commented on cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll:
```diff
@@ -41,7 +41,9 @@ predicate ignoreInstruction(Instruction instr) {
     instr instanceof ReturnIndirectionInstruction or
-    instr instanceof UninitializedGroupInstruction
+    instr instanceof UninitializedGroupInstruction or
+    // We exclude instruction belonging to functions that have a summary.
```
```suggestion
    // We exclude instructions belonging to functions that have a summary.
```

jketema reviewed (commented):
I guess this sort of assumes that our modelling for functions with summaries is in some sense "complete" (we didn't miss modelling any of the possible flow)?

What's going on with the significant reduction in result tuples on Samate for `cpp/uncontrolled-allocation-size`?

aschackmull commented:
Please compare with other languages. This is something we've tackled quite a bit already. The basic intent is of course that for callables with good models we want to use flow through those summaries and not the source, but there's a whole range of complexity here.
Firstly, I believe the general consensus so far has been, not to exclude bodies from the `DataFlow::Node` type, but rather to adjust the dispatch relation. I.e. for a callable with a good model, we should only dispatch to the model summary and not the  source. This means that we'll still be able to make use of `DataFlow::Node`s within that callable body for other use-cases, and we won't get a FN if, say, there's a flow-based alert contained within that body (not making use of dispatch to/from the callable).
Then there's the added complexity around when you actually want to do this. Here's the situation for Java:
* Models have provenance that affect our confidence in them: they can be manual or generated.
* Models can be exact or not relative to a given call: if the static call target matches the type in the yaml then it's exact, and if it targets an override, then it's not exact.
* Exact manual models have the highest priority, they always suppress source dispatch.
* Source dispatch targets are given a heuristic confidence, either high or low depending on the declaring type of the static target (e.g. JDK interfaces like List and Map are low confidence).
* Manual models suppress low-confidence source dispatch.
* Finally, generated models are suppressed in those cases where the static call target is in the analysed source (in order to make model generation idempotent).
* For all other cases we dispatch to both any appliccable model and any potential source target. I.e. generated models never suppress source dispatch, and manual models only suppress source dispatch when they are either exact or the source dispatch is low confidence.

(see https://github.com/github/codeql/blob/main/java/ql/lib/semmle/code/java/dataflow/internal/DataFlowDispatch.qll)

aschackmull commented:
cc @michaelnebel Could you read through the above situation for Java and list differences to C#, if any?

michaelnebel commented:
> cc @michaelnebel Could you read through the above situation for Java and list differences to C#, if any?

Sure, I will try and provide an overview
For C# the logic is a little less fine grained.
- Models have provenance that affect our confidence in them: they can be manual or generated (same as for Java).
- There is no concept of *exact*.
- There is no concept of low or high confidence source dispatch. If a manual model exist for a callable, both the summarized callable and the (source) callable will be used as data flow dispatch targets.
- A summarized callable based on a generated summary will only be used in case there is no other possible target in source (for C# we consider both static and dynamic targets - maybe this needs to adjusted to only consider static targets like for Java).
- Neutral models do not suppress any call targets (contrary to java). However, neutral models impact how flow is propagated via a summarized callables based on generated summaries.

MathiasVP commented:
Thanks for all the comments, Schack! I'm aware of the complexities in Java and C# around this here.

C++ is currently in a different situation, however:
- We haven't done any autogeneration of models yet. So all of our MaD models are manually generated precise models
- C++ has very little focus on dispatch. Basically, all of our interesting calls are either function pointers to user-defined defined functions, or they're static calls. Thus, we don't really need to care about whether they target an override or not.
- The way people write generic code in C++ that works on a arbitrary list or map container is by using _templates_, and since all our analysis happens on template instantiations it's my current understanding that we don't have the same problem map or lists as Java does.

MathiasVP commented:
Oh, and thanks Michael! I'll make sure to record all this in an issue somewhere so that it's easily accessible if C++ ever gets to a Glorious model generation Future 🙂

MathiasVP commented:
> I guess this sort of assumes that our modelling for functions with summaries is in some sense "complete" (we didn't miss modelling any of the possible flow)?

Correct. Luckily, I believe that's currently the case. For example, we haven't added MaD summaries for `std::map` since we can't model them precisely yet (and thus we stick to the flow generated from analysing the source code directly).

> What's going on with the significant reduction in result tuples on Samate for `cpp/uncontrolled-allocation-size`?

Good catch. I missed that table. I'll take a look

aschackmull commented:
With this background in place, I think the way forward for C++ ought to be:
* Don't remove `DataFlow::Node`s.
* Instead: remove call edges to source functions that have a MaD summary, i.e. use the summary and suppress the source dispatch.

MathiasVP commented:
Removing call edges to the source function sounds like a fine alternative, indeed.

I'm still not convinced we gain anything by generating those dataflow nodes. I've never seen a useful alert inside the implementation of an STL container so I don't see what we gain by having those. I'll discuss this with the C team offline and get back with the conclusions here

MathiasVP commented:
Oh well. I didn't manage to start any interesting discussions offline, but in order to get something merged that's less controversial I've gone with @aschackmull's suggestion and removed the the call edges to the source function when there's a MaD summary.

Note that this above isn't enough to completely get rid of source-provided flow: 4e44201ba8ecbca41b3611d4677de95aa23203c1 shows that we regain the alert in `test.cpp` when all we do is remove those call edges. The reason is that we still get flow-through from a source parameter to a summarized function (and we also lose the new `cpp/path-injection` as well since the field-flow branch limit is once again hit). The 83dcaae4c2651c4a697ae6c6bd37b29b44a00c32 is to remove those parameter nodes as well, and this brings us back to the same set of results as before.

I was about to suggest that we could've also removed the source `ReturnNode`s, but that would still allow summary flow through `TParamReturnNode`. So I think removing the parameter nodes is the best thing to do.

jketema commented:
> Note that this above isn't enough to completely get rid of source-provided flow: [4e44201](https://github.com/github/codeql/commit/4e44201ba8ecbca41b3611d4677de95aa23203c1) shows that we regain the alert in `test.cpp` when all we do is remove those call edges. The reason is that we still get flow-through from a source parameter to a summarized function (and we also lose the new `cpp/path-injection` as well since the field-flow branch limit is once again hit). The [83dcaae](https://github.com/github/codeql/commit/83dcaae4c2651c4a697ae6c6bd37b29b44a00c32) is to remove those parameter nodes as well, and this brings us back to the same set of results as before.

Shouldn't we just removing edges here too?

aschackmull commented:
> > Note that this above isn't enough to completely get rid of source-provided flow: [4e44201](https://github.com/github/codeql/commit/4e44201ba8ecbca41b3611d4677de95aa23203c1) shows that we regain the alert in `test.cpp` when all we do is remove those call edges. The reason is that we still get flow-through from a source parameter to a summarized function (and we also lose the new `cpp/path-injection` as well since the field-flow branch limit is once again hit). The [83dcaae](https://github.com/github/codeql/commit/83dcaae4c2651c4a697ae6c6bd37b29b44a00c32) is to remove those parameter nodes as well, and this brings us back to the same set of results as before.
> 
> Shouldn't we just removing edges here too?

Yeah, something is off here. I don't really know what's going on, but if the existence of the parameter nodes are affecting the test, then I don't think the call edges are being correctly removed in the call graph given to data flow.

MathiasVP commented:
I'm writing up the discussion offline right now. Stay tuned!

MathiasVP commented:
I had a good discussion with Schack about what was going on and we found quite a few MaD issues. Interestingly, they all kind of "aligned" so that no consistency issues were being reported since everything was consistent while still being wrong 😂

The bugs are as follows:
1. We had chosen (due to misunderstandings when we introduced MaD more than a year ago) to skip generating `DataFlowCallable`s when we had a summary for the function. This is totally not what we're supposed to do, and this has been fixed in 662e74924b15e34b5d16fe182943664286acef8e
2. We were conflating the enclosing callable when there was both a summary and a source function. Due to bug number 1 this was totally fine, but as soon as we reintroduced those missing callables in 662e74924b15e34b5d16fe182943664286acef8e this made all kinds of consistency errors go crazy.  This has been fixed in d40322f9eb8ffb5c84e8577efa638a7796dd4c4b.
3. Finally, the remapping of parameters I did in de3b111f2f5c1f5cb69292c0bcd0e77b2f5ce5a6 now also turned to be problematic when the missing callables were introduced. This has been fixed in 06bc8add9d68ce8137f13d5d0b968d8ea4f2896f

Thanks for all the help on this, @aschackmull! I believe the PR is good to go now 🤞 I'll run another DCA just to make sure the new good result is still there

aschackmull reviewed (approved):
Looks like it matches the changes we went through together, so 👍 from me.

MathiasVP commented:
Bahh... We're no longer getting that new `cpp/path-injection` result 😭 So some change caused us to go above the field-flow branch limit again. I'll investigate!

So while this PR doesn't actually solve what I initially wanted to solve (i.e., we were going over the field-flow branch limit unnecessarily) I think it ended up fixing a lot of things.

DCA looks fine other than that, so I'll merge this and investigate why we're no longer gaining that result.
